app/messages/service_credential_app_binding_create_message.rb

```diff
@@ -0,0 +1,20 @@
+module VCAP::CloudController
+  class ServiceCredentialAppBindingCreateMessage < ServiceCredentialBindingCreateMessage
```

Comment:
I find this a bit confusing. The base class validates that the type is either 'app' or 'key' and then the subclass can only be one of those types. Also not a big fan of having `if type==app` scattered throughout the controller. I am wondering if we just should create a simple message to extract the type and then take completely different paths depending on that. So we can create a specific message and do any validations/action calls that we need from them on thinking just on one kind of resource. There is no much reuse of code provided by the inheritance in the messages to actually care too much for it.

Reply:
The reason I like the reduplication here is not about the code repetition but because these messages should always be kept in sync and are represented as the same thing to API client. E.g a future change on SI relationship should affect both key and app type, hence it makes sense to keep the SI extracted. 
I think this is not the perfect pattern, but it is following the same pattern as the service instance create message and I think at this point in the project it might be more beneficial to have similar patterns between al endpoint.